### ISSUE TITLE
Break up parser over multiple modules

### ISSUE THREAD

**Issue by tmjoen** [markglenn/liquex#18]
From `nimble_parsec`'s CHANGELOG:

```
Add defcombinator and support for remote parsec. 
This allows larger parsers to be broken over multiple modules 
to speed up compilation by leveraging parallelism.
```

Is this something that could be leveraged by Liquex? 

Compiling takes quite a bit of time (naturally, since it's a pretty big parser) -- and when I extend `base_element` to bring in my own custom elements, it will compile liquex first and then my own extended parser afterwards (effectively compiling much of the same code twice). There is maybe no way around this?

Sorry for raising an issue, I was just curious if there was anything to be done here!

Thanks again for a lovely library <3

**Comment by markglenn** [comment]
Thank you for bringing this up. I've been experimenting with different options to help improve the compilation speeds, but haven't found anything as of yet. I didn't use defcombinator more than at the top level for the document and now literals (to solve a recent recursion issue), so I may try breaking down other large pieces into their own combinators.

Thank you for the tip, and I'll try it out.

**Comment by markglenn** [comment]
> The only situation where you should use parsec/2 for composition is when a large parser is used over and over again in a way compilation times are high. In this sense, you can use parsec/2 to improve compilation time at the cost of runtime performance. By using parsec/2, the tree size built at compile time will be reduced although runtime performance is degraded as every time this function is invoked it introduces a stacktrace entry.

It appears that there is a runtime cost to breaking up the parser, so I will have to balance that with compilation costs.

**Comment by tmjoen** [comment]
Ah, I missed that. I'd rather live with a bit longer compilation time than a runtime penalty, personally!

My parser implementation very rarely changes, so maybe I'd have a bit more luck with https://hexdocs.pm/nimble_parsec/Mix.Tasks.NimbleParsec.Compile.html#module-examples

**Comment by markglenn** [comment]
I found a large speedup around argument parsing.  Because argument parsing had become recursive in #15, this was already split off into its own combinator.  By using this combinator every time the `argument/1` function was used, I was able to reduce my compilation time from 15 seconds down to under 4 seconds.

I don't feel like this will negatively affect the runtime cost in any noticeable way, and possibly fixes other issues where the arguments were not being properly parsed in a recursive manner.
